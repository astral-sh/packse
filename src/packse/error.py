import textwrap
from pathlib import Path


class PackseError(Exception):
    """Base type for all Packse exceptions"""


class UserError(PackseError):
    """It's the user's fault :)"""


class RequiresExtra(UserError):
    def __init__(self, feature: str, extra: str) -> None:
        message = f"Using {feature} requires installation with extra {extra!r}"
        super().__init__(message)


class DestinationAlreadyExists(UserError):
    def __init__(self, destination: Path) -> None:
        message = f"Destination directory '{destination}' already exists"
        super().__init__(message)


class InvalidScenario(UserError):
    def __init__(self, path: Path, reason: str) -> None:
        message = f"File at '{path}' is not a valid scenario: {reason}"
        super().__init__(message)


class InvalidPackageVersion(UserError):
    """Version is not PEP compliant"""

    def __init__(self, version: str) -> None:
        message = f"Version {version!r} is not valid."
        super().__init__(message)


class FileNotFound(UserError):
    def __init__(self, path: Path) -> None:
        message = f"File '{path}' not found"
        super().__init__(message)


class BuildError(PackseError):
    def __init__(self, message, output) -> None:
        message = message + ":\n" + textwrap.indent(output, " " * 4)
        super().__init__(message)


class PublishError(PackseError):
    pass


class PublishToolError(PublishError):
    def __init__(self, message, output) -> None:
        message = message + ":\n" + textwrap.indent(output, " " * 4)
        super().__init__(message)


class PublishNoCredentials(PublishError):
    def __init__(self) -> None:
        message = (
            "No credentials found for publish!\n"
            "Provide an API token via `PACKSE_PUBLISH_PASSWORD` or disable authentication with `--anonymous`"
        )
        super().__init__(message)


class PublishAlreadyExists(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish for '{package}' already exists"
        super().__init__(message)


class PublishRateLimit(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish of '{package}' failed due to rate limits"
        super().__init__(message)


class PublishConnectionError(PublishError):
    def __init__(self, package: str) -> None:
        message = f"Publish of '{package}' failed due to a connection error"
        super().__init__(message)


class InvalidPublishTarget(UserError):
    def __init__(self, path: Path, reason: str) -> None:
        message = f"Publish target at '{path}' is not a valid: {reason}"
        super().__init__(message)


class ServeError(PackseError):
    def __init__(self, message: str) -> None:
        super().__init__(message)


class ServeThreadError(ServeError):
    def __init__(self, message: str) -> None:
        super().__init__(message)


class ServeCommandError(ServeError):
    def __init__(self, message: str, stderr: str) -> None:
        if stderr.strip():
            message += f":\n\n{stderr}"
        super().__init__(message)


class ServeAddressInUse(ServeError):
    def __init__(self, url: str) -> None:
        message = f"Address '{url}' already in use"
        super().__init__(message)


class ServeAlreadyRunning(ServeError):
    def __init__(self) -> None:
        message = "Server is already running"
        super().__init__(message)
