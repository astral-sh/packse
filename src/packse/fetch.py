import logging
import shutil
import subprocess
import tempfile
import time
from pathlib import Path

import pkg_resources

from packse.error import DestinationAlreadyExists

logger = logging.getLogger(__name__)


def fetch(
    dest: Path | None = None,
    ref: str | None = None,
    repo_url: str = "https://github.com/astral-sh/packse",
    repo_subdir: str = "scenarios",
    force: bool = False,
):
    start_time = time.time()
    dest = dest or (Path.cwd() / "scenarios")

    if dest.exists() and not force:
        raise DestinationAlreadyExists(dest)

    if not ref:
        ref = pkg_resources.get_distribution("packse").version
        if ref == "0.0.0":
            ref = "HEAD"

    with tempfile.TemporaryDirectory() as tmpdir:
        # Perform a sparse checkout where we only grab the `scenarios` folder
        logger.info("Cloning repository %s", repo_url)
        subprocess.check_call(
            [
                "git",
                "clone",
                "-n",
                "--filter=tree:0",
                repo_url,
                "repo",
            ],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=tmpdir,
        )

        logger.info("Checking out directory '%s' at ref %s", repo_subdir, ref)
        subprocess.check_call(
            ["git", "sparse-checkout", "set", "--no-cone", repo_subdir],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )
        subprocess.check_call(
            ["git", "fetch", "origin", ref],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )
        subprocess.check_call(
            ["git", "checkout", ref],
            stdout=subprocess.DEVNULL,
            stderr=subprocess.STDOUT,
            cwd=Path(tmpdir) / "repo",
        )

        src = Path(tmpdir) / "repo" / repo_subdir
        file_count = 0
        for file in sorted(src.iterdir()):
            file_count += 1
            logger.info("Found %s", file.name)

        logger.info("Copying files into '%s'", dest)
        shutil.copytree(src, dest, dirs_exist_ok=True)

    logger.info(
        "Fetched %s files in %.2fs",
        file_count,
        time.time() - start_time,
    )
