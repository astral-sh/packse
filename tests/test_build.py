import os
from pathlib import Path

import pytest
from packse import __development_base_path__
from packse.scenario import load_scenario, scenario_hash

from .common import snapshot_command


def test_build_no_target(snapshot):
    assert snapshot_command(["build"]) == snapshot


def test_build_target_does_not_exist(snapshot):
    assert snapshot_command(["build", "foo"]) == snapshot


def test_build_invalid_target(snapshot, tmpcwd):
    target = tmpcwd / "test.json"
    target.touch()
    assert snapshot_command(["build", str(target)]) == snapshot


@pytest.mark.usefixtures("tmpcwd")
def test_build_example(snapshot):
    target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    assert (
        snapshot_command(
            ["build", str(target)], snapshot_filesystem=True, snapshot_stderr=False
        )
        == snapshot
    )


@pytest.mark.usefixtures("tmpcwd")
def test_build_example_short_names(snapshot):
    target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    assert (
        snapshot_command(
            ["build", str(target), "--short-names"],
            snapshot_filesystem=True,
            snapshot_stderr=False,
        )
        == snapshot
    )


@pytest.mark.usefixtures("tmpcwd")
@pytest.mark.usefixtures("tmpenviron")
def test_build_example_with_seed(snapshot):
    target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    os.environ["PACKSE_VERSION_SEED"] = "foo"
    assert (
        snapshot_command(
            ["build", str(target)], snapshot_filesystem=True, snapshot_stderr=False
        )
        == snapshot
    )


@pytest.mark.usefixtures("tmpcwd")
def test_build_example_already_exists(snapshot):
    target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    scenario = load_scenario(target)
    name = f"{scenario.name}-{scenario_hash(scenario)}"
    (Path.cwd() / "build" / name).mkdir(parents=True)
    assert snapshot_command(["build", target], snapshot_filesystem=True) == snapshot


@pytest.mark.usefixtures("tmpcwd")
def test_build_example_already_exists_with_rm_flag(snapshot):
    target = __development_base_path__ / "scenarios" / "examples" / "example.json"
    (Path(".") / "build").mkdir()
    assert (
        snapshot_command(["build", target, "--rm"], snapshot_stderr=False) == snapshot
    )
