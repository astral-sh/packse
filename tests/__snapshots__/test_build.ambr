# serializer version: 1
# name: test_build_example
  dict({
    'exit_code': 0,
    'filesystem': dict({
      'build/example-961b4c22/example-961b4c22-0.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/example-961b4c22-0.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_961b4c22"]
        
        [project]
        name = "example-961b4c22"
        version = "0.0.0"
        dependencies = [
            '''example-a-961b4c22''',
        ]
        requires-python = ">=3.8"
        description = '''This is an example scenario, in which the user depends on a single package `a` which requires `b`.'''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/example-961b4c22-0.0.0/src/example_961b4c22/__init__.py': '''
        __version__ = "0.0.0"
  
      ''',
      'build/example-961b4c22/example-a-961b4c22-1.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/example-a-961b4c22-1.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_a_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_a_961b4c22"]
        
        [project]
        name = "example-a-961b4c22"
        version = "1.0.0"
        dependencies = [
            '''example-b-961b4c22>1.0.0''',
        ]
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/example-a-961b4c22-1.0.0/src/example_a_961b4c22/__init__.py': '''
        __version__ = "1.0.0"
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-1.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-1.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_b_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_b_961b4c22"]
        
        [project]
        name = "example-b-961b4c22"
        version = "1.0.0"
        dependencies = []
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-1.0.0/src/example_b_961b4c22/__init__.py': '''
        __version__ = "1.0.0"
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-2.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-2.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_b_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_b_961b4c22"]
        
        [project]
        name = "example-b-961b4c22"
        version = "2.0.0"
        dependencies = []
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-2.0.0/src/example_b_961b4c22/__init__.py': '''
        __version__ = "2.0.0"
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-3.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-3.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_b_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_b_961b4c22"]
        
        [project]
        name = "example-b-961b4c22"
        version = "3.0.0"
        dependencies = [
            '''example-c-961b4c22''',
        ]
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/example-b-961b4c22-3.0.0/src/example_b_961b4c22/__init__.py': '''
        __version__ = "3.0.0"
  
      ''',
      'dist/example-961b4c22/example_961b4c22-0.0.0.tar.gz': 'md5:78387c54dc1123fc169bde1ad63aa11b',
      'dist/example-961b4c22/example_a_961b4c22-1.0.0-py3-none-any.whl': 'md5:93f87d75d52b4dcac32ea670f9c7c5ab',
      'dist/example-961b4c22/example_a_961b4c22-1.0.0.tar.gz': 'md5:8ed996db89e98df9b698f9e90bec5038',
      'dist/example-961b4c22/example_b_961b4c22-1.0.0-py3-none-any.whl': 'md5:96d44ceacde16b72625ffbed26c04377',
      'dist/example-961b4c22/example_b_961b4c22-1.0.0.tar.gz': 'md5:6a6fe7b8bb871d712dc7883dc0cb0d68',
      'dist/example-961b4c22/example_b_961b4c22-2.0.0-py3-none-any.whl': 'md5:fb9ab8eb91c87cf924ce14e3ed074ce0',
      'dist/example-961b4c22/example_b_961b4c22-2.0.0.tar.gz': 'md5:82b9995b2b3738d5975416105d1ab986',
      'dist/example-961b4c22/example_b_961b4c22-3.0.0-py3-none-any.whl': 'md5:3b8a4c8aa5676269e8348b747178b381',
      'dist/example-961b4c22/example_b_961b4c22-3.0.0.tar.gz': 'md5:56f1ef2dedf69f360c55297e9d9b1f20',
      'tree': '''
        test_build_example0
        ├── build
        │   └── example-961b4c22
        │       ├── example-961b4c22-0.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_961b4c22
        │       │           └── __init__.py
        │       ├── example-a-961b4c22-1.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_a_961b4c22
        │       │           └── __init__.py
        │       ├── example-b-961b4c22-1.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_b_961b4c22
        │       │           └── __init__.py
        │       ├── example-b-961b4c22-2.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_b_961b4c22
        │       │           └── __init__.py
        │       └── example-b-961b4c22-3.0.0
        │           ├── .gitignore
        │           ├── pyproject.toml
        │           └── src
        │               └── example_b_961b4c22
        │                   └── __init__.py
        └── dist
            └── example-961b4c22
                ├── example_961b4c22-0.0.0.tar.gz
                ├── example_a_961b4c22-1.0.0-py3-none-any.whl
                ├── example_a_961b4c22-1.0.0.tar.gz
                ├── example_b_961b4c22-1.0.0-py3-none-any.whl
                ├── example_b_961b4c22-1.0.0.tar.gz
                ├── example_b_961b4c22-2.0.0-py3-none-any.whl
                ├── example_b_961b4c22-2.0.0.tar.gz
                ├── example_b_961b4c22-3.0.0-py3-none-any.whl
                └── example_b_961b4c22-3.0.0.tar.gz
        
        19 directories, 24 files
  
      ''',
    }),
    'stderr': '<not included>',
    'stdout': '''
      example-961b4c22
  
    ''',
  })
# ---
# name: test_build_example_already_exists
  dict({
    'exit_code': 1,
    'filesystem': dict({
      'tree': '''
        test_build_example_already_exi0
        └── build
            └── example-961b4c22
        
        2 directories
  
      ''',
    }),
    'stderr': '''
      Building 'example-961b4c22' in directory '[PWD]/build/example-961b4c22'
      Destination directory '[PWD]/build/example-961b4c22' already exists. Pass `--force` to allow removal.
  
    ''',
    'stdout': '',
  })
# ---
# name: test_build_example_already_exists_with_rm_flag
  dict({
    'exit_code': 2,
    'stderr': '<not included>',
    'stdout': '',
  })
# ---
# name: test_build_example_short_names
  dict({
    'exit_code': 0,
    'filesystem': dict({
      'build/example-961b4c22/a-961b4c22-1.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/a-961b4c22-1.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/a_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/a_961b4c22"]
        
        [project]
        name = "a-961b4c22"
        version = "1.0.0"
        dependencies = [
            '''b-961b4c22>1.0.0''',
        ]
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/a-961b4c22-1.0.0/src/a_961b4c22/__init__.py': '''
        __version__ = "1.0.0"
  
      ''',
      'build/example-961b4c22/b-961b4c22-1.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/b-961b4c22-1.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/b_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/b_961b4c22"]
        
        [project]
        name = "b-961b4c22"
        version = "1.0.0"
        dependencies = []
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/b-961b4c22-1.0.0/src/b_961b4c22/__init__.py': '''
        __version__ = "1.0.0"
  
      ''',
      'build/example-961b4c22/b-961b4c22-2.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/b-961b4c22-2.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/b_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/b_961b4c22"]
        
        [project]
        name = "b-961b4c22"
        version = "2.0.0"
        dependencies = []
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/b-961b4c22-2.0.0/src/b_961b4c22/__init__.py': '''
        __version__ = "2.0.0"
  
      ''',
      'build/example-961b4c22/b-961b4c22-3.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/b-961b4c22-3.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/b_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/b_961b4c22"]
        
        [project]
        name = "b-961b4c22"
        version = "3.0.0"
        dependencies = [
            '''c-961b4c22''',
        ]
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/b-961b4c22-3.0.0/src/b_961b4c22/__init__.py': '''
        __version__ = "3.0.0"
  
      ''',
      'build/example-961b4c22/example-961b4c22-0.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-961b4c22/example-961b4c22-0.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_961b4c22"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_961b4c22"]
        
        [project]
        name = "example-961b4c22"
        version = "0.0.0"
        dependencies = [
            '''a-961b4c22''',
        ]
        requires-python = ">=3.8"
        description = '''This is an example scenario, in which the user depends on a single package `a` which requires `b`.'''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-961b4c22/example-961b4c22-0.0.0/src/example_961b4c22/__init__.py': '''
        __version__ = "0.0.0"
  
      ''',
      'dist/example-961b4c22/a_961b4c22-1.0.0-py3-none-any.whl': 'md5:51bf95659901e5f92e347d24ccb6b1e2',
      'dist/example-961b4c22/a_961b4c22-1.0.0.tar.gz': 'md5:82fbb6a7e47cb65004d7e1ed960e571a',
      'dist/example-961b4c22/b_961b4c22-1.0.0-py3-none-any.whl': 'md5:26e16db87bf22e82aeca0810ef81d01b',
      'dist/example-961b4c22/b_961b4c22-1.0.0.tar.gz': 'md5:b0c6c32380822d4fd18549079586737e',
      'dist/example-961b4c22/b_961b4c22-2.0.0-py3-none-any.whl': 'md5:c351f0481dd14bf3d79b063acd15cb40',
      'dist/example-961b4c22/b_961b4c22-2.0.0.tar.gz': 'md5:60eb62ac37488a6c4106b153b955b9c8',
      'dist/example-961b4c22/b_961b4c22-3.0.0-py3-none-any.whl': 'md5:2bcb41471376e19cf5732ec12fd7c4c5',
      'dist/example-961b4c22/b_961b4c22-3.0.0.tar.gz': 'md5:f716df3ed041ab73dacc064313708f02',
      'dist/example-961b4c22/example_961b4c22-0.0.0.tar.gz': 'md5:770dff7d79ec4d77b36c632cafa3a056',
      'tree': '''
        test_build_example_short_names0
        ├── build
        │   └── example-961b4c22
        │       ├── a-961b4c22-1.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── a_961b4c22
        │       │           └── __init__.py
        │       ├── b-961b4c22-1.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── b_961b4c22
        │       │           └── __init__.py
        │       ├── b-961b4c22-2.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── b_961b4c22
        │       │           └── __init__.py
        │       ├── b-961b4c22-3.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── b_961b4c22
        │       │           └── __init__.py
        │       └── example-961b4c22-0.0.0
        │           ├── .gitignore
        │           ├── pyproject.toml
        │           └── src
        │               └── example_961b4c22
        │                   └── __init__.py
        └── dist
            └── example-961b4c22
                ├── a_961b4c22-1.0.0-py3-none-any.whl
                ├── a_961b4c22-1.0.0.tar.gz
                ├── b_961b4c22-1.0.0-py3-none-any.whl
                ├── b_961b4c22-1.0.0.tar.gz
                ├── b_961b4c22-2.0.0-py3-none-any.whl
                ├── b_961b4c22-2.0.0.tar.gz
                ├── b_961b4c22-3.0.0-py3-none-any.whl
                ├── b_961b4c22-3.0.0.tar.gz
                └── example_961b4c22-0.0.0.tar.gz
        
        19 directories, 24 files
  
      ''',
    }),
    'stderr': '<not included>',
    'stdout': '''
      example-961b4c22
  
    ''',
  })
# ---
# name: test_build_example_with_seed
  dict({
    'exit_code': 0,
    'filesystem': dict({
      'build/example-7661fdb8/example-7661fdb8-0.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-7661fdb8/example-7661fdb8-0.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_7661fdb8"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_7661fdb8"]
        
        [project]
        name = "example-7661fdb8"
        version = "0.0.0"
        dependencies = [
            '''example-a-7661fdb8''',
        ]
        requires-python = ">=3.8"
        description = '''This is an example scenario, in which the user depends on a single package `a` which requires `b`.'''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-7661fdb8/example-7661fdb8-0.0.0/src/example_7661fdb8/__init__.py': '''
        __version__ = "0.0.0"
  
      ''',
      'build/example-7661fdb8/example-a-7661fdb8-1.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-7661fdb8/example-a-7661fdb8-1.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_a_7661fdb8"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_a_7661fdb8"]
        
        [project]
        name = "example-a-7661fdb8"
        version = "1.0.0"
        dependencies = [
            '''example-b-7661fdb8>1.0.0''',
        ]
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-7661fdb8/example-a-7661fdb8-1.0.0/src/example_a_7661fdb8/__init__.py': '''
        __version__ = "1.0.0"
  
      ''',
      'build/example-7661fdb8/example-b-7661fdb8-1.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-7661fdb8/example-b-7661fdb8-1.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_b_7661fdb8"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_b_7661fdb8"]
        
        [project]
        name = "example-b-7661fdb8"
        version = "1.0.0"
        dependencies = []
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-7661fdb8/example-b-7661fdb8-1.0.0/src/example_b_7661fdb8/__init__.py': '''
        __version__ = "1.0.0"
  
      ''',
      'build/example-7661fdb8/example-b-7661fdb8-2.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-7661fdb8/example-b-7661fdb8-2.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_b_7661fdb8"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_b_7661fdb8"]
        
        [project]
        name = "example-b-7661fdb8"
        version = "2.0.0"
        dependencies = []
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-7661fdb8/example-b-7661fdb8-2.0.0/src/example_b_7661fdb8/__init__.py': '''
        __version__ = "2.0.0"
  
      ''',
      'build/example-7661fdb8/example-b-7661fdb8-3.0.0/.gitignore': '''
        __pycache__
        *.pyc
  
      ''',
      'build/example-7661fdb8/example-b-7661fdb8-3.0.0/pyproject.toml': '''
        [build-system]
        requires = ["hatchling"]
        build-backend = "hatchling.build"
        
        [tool.hatch.build.targets.wheel]
        packages = ["src/example_b_7661fdb8"]
        
        [tool.hatch.build.targets.sdist]
        only-include = ["src/example_b_7661fdb8"]
        
        [project]
        name = "example-b-7661fdb8"
        version = "3.0.0"
        dependencies = [
            '''example-c-7661fdb8''',
        ]
        requires-python = ">=3.8"
        description = ''''''
        
        [project.optional-dependencies]
  
      ''',
      'build/example-7661fdb8/example-b-7661fdb8-3.0.0/src/example_b_7661fdb8/__init__.py': '''
        __version__ = "3.0.0"
  
      ''',
      'dist/example-7661fdb8/example_7661fdb8-0.0.0.tar.gz': 'md5:028f372c7482f37edfbc8490f89d881e',
      'dist/example-7661fdb8/example_a_7661fdb8-1.0.0-py3-none-any.whl': 'md5:fa6f4b9b486012c02db2e99fe86583d8',
      'dist/example-7661fdb8/example_a_7661fdb8-1.0.0.tar.gz': 'md5:8df15e04d35c06db3f5e5aa940a95a4e',
      'dist/example-7661fdb8/example_b_7661fdb8-1.0.0-py3-none-any.whl': 'md5:7b1f7e1df5cc98bd0486070555bf2d27',
      'dist/example-7661fdb8/example_b_7661fdb8-1.0.0.tar.gz': 'md5:8c415e5939a6a707046799465c67fda9',
      'dist/example-7661fdb8/example_b_7661fdb8-2.0.0-py3-none-any.whl': 'md5:419d8c7d887a31d74a063c8e85479a4c',
      'dist/example-7661fdb8/example_b_7661fdb8-2.0.0.tar.gz': 'md5:5c44cfb26db7c07c846c1a1dc39aeeda',
      'dist/example-7661fdb8/example_b_7661fdb8-3.0.0-py3-none-any.whl': 'md5:770c148a40e04e526ba886874f20d9a3',
      'dist/example-7661fdb8/example_b_7661fdb8-3.0.0.tar.gz': 'md5:0294ccfa865be65ea82ee929dcb27b70',
      'tree': '''
        test_build_example_with_seed0
        ├── build
        │   └── example-7661fdb8
        │       ├── example-7661fdb8-0.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_7661fdb8
        │       │           └── __init__.py
        │       ├── example-a-7661fdb8-1.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_a_7661fdb8
        │       │           └── __init__.py
        │       ├── example-b-7661fdb8-1.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_b_7661fdb8
        │       │           └── __init__.py
        │       ├── example-b-7661fdb8-2.0.0
        │       │   ├── .gitignore
        │       │   ├── pyproject.toml
        │       │   └── src
        │       │       └── example_b_7661fdb8
        │       │           └── __init__.py
        │       └── example-b-7661fdb8-3.0.0
        │           ├── .gitignore
        │           ├── pyproject.toml
        │           └── src
        │               └── example_b_7661fdb8
        │                   └── __init__.py
        └── dist
            └── example-7661fdb8
                ├── example_7661fdb8-0.0.0.tar.gz
                ├── example_a_7661fdb8-1.0.0-py3-none-any.whl
                ├── example_a_7661fdb8-1.0.0.tar.gz
                ├── example_b_7661fdb8-1.0.0-py3-none-any.whl
                ├── example_b_7661fdb8-1.0.0.tar.gz
                ├── example_b_7661fdb8-2.0.0-py3-none-any.whl
                ├── example_b_7661fdb8-2.0.0.tar.gz
                ├── example_b_7661fdb8-3.0.0-py3-none-any.whl
                └── example_b_7661fdb8-3.0.0.tar.gz
        
        19 directories, 24 files
  
      ''',
    }),
    'stderr': '<not included>',
    'stdout': '''
      example-7661fdb8
  
    ''',
  })
# ---
# name: test_build_invalid_target
  dict({
    'exit_code': 1,
    'stderr': '''
      File at '[PWD]/test.json' is not a valid scenario: Input data was truncated.
  
    ''',
    'stdout': '',
  })
# ---
# name: test_build_no_target
  dict({
    'exit_code': 2,
    'stderr': '''
      usage: packse build [-h] [--force] [--short-names] [--no-hash] [--skip-root]
                          [-v] [-q]
                          targets [targets ...]
      packse build: error: the following arguments are required: targets
  
    ''',
    'stdout': '',
  })
# ---
# name: test_build_target_does_not_exist
  dict({
    'exit_code': 1,
    'stderr': '''
      File 'foo' not found.
  
    ''',
    'stdout': '',
  })
# ---
